docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,679 @@
+
+  The analyzer would still need to separately encode the fact that `"RuntimeFeature.IsDynamicCodeSupported"` corresponds to `RequiresDynamicCodeAttribute`.
+
+- a reference to the feature attribute, for those feature switches that are associated with attributes:
+
+  ```csharp
+  class RuntimeFeature {
+      [FeatureSwitch(typeof(RequiresDynamicCodeAttribute))]
+      public static bool IsDynamicCodeSupported => AppContext.TryGetSwitch("RuntimteFeature.IsDynamicCodeSupported", out bool isEnabled) ? isEnabled : true;
+  }
+  ```
+
+  In this case, ILCompiler would need to hard-code the fact that `RequiresDynamicCodeAttribute` corresponds to `"RuntimeFeature.IsDynamicCodeSupported"`, and use this knowledge to treat the property as returning a constant.
+
+  However, the Roslyn analyzer would have enough information from this attribute alone.
+
+## Relationship between feature switches and feature guards
+
+### Feature switches are also guards
+A feature switch is also a feature guard for the feature it is defined by. We could encode this in one of two ways:
+- `FeatureSwitch` only, with a unified representation that includes the mapping to both the attribute and feature name, or
+- Require both `FeatureSwitch` and `FeatureGuard`, where both attributes together provide a mapping to the attribute and the feature name
+
+### Feature guards _may_ also be feature switches
+
+A feature guard might also come with its own independent feature switch. We shoud be careful to avoid violating the assumptions of the feature guard by controlling the property via a feature switch.
+
+For example, if we had a separate feature switch for `IsDynamicCodeCompiled`, this would allow setting `IsDynamicCodeCompiled` to `true` even when `IsDynamicCodeSupported` is `false`. This is essentially what we did for features like `StartupHookSupport`: this feature switch can be set even in trimmed apps.
+
+We rely on trim warnings to alert the app author to the problem, and also default `StartupHookSupport` to `false` for trimmed apps. If we have an attribute-based model for feature guards, we may want to consider inferring these defaults from the `FeatureGuard` instead (so a guard for `RequiresUnreferencedCode` that is also a feature switch would be `false` by default whenever "unreferenced code" is unavailable).
+
+The proposal for now is not to infer any defaults and just take care to set appropriate defaults in the SDK. This means that custom feature guards that are also feature switches will need to do the same. For example, a library that defines a feature switch which guards `RequiresUnreferencedCode` will need to ship with MSBuild targets that disable the feature by default for trimmed apps.
```

Comment:
I think we should add a sample for this - this part is probably the most complex topic in this document. Would be great to have a sample over startup hook, or something similar. How you tie a guard to a different feature switch and how it's supposed to work.

The defaults discussions is actually only important for illink/ilcompiler. The analyzer shouldn't really care as it needs to analyze both branches anyway. Also the correctness of this is not a problem because the "true" branch were the warnings are suppressed would still produce warnings in the app (illink/ilcompiler) if the feature switch is enabled. We basically need to be in a place where analyzer doesn't produce warnings in any case, but trimmer will produce them depending on the feature switch value.